### PR TITLE
Enable context parallelism for RL training

### PR DESCRIPTION
**Pull request by tdene**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/rl/rl_utils.py [hunk 1]
```diff
@@ -19,6 +19,12 @@
 import numpy as np
 import torch
 import torch.distributed as dist
+import torch.distributed.nn
+
+try:
+    import transformer_engine_torch as tex
+except ImportError:
+    tex = None
```
Comment (yobibyte) [inline megatron/rl/rl_utils.py]: I would use HAVE_TEX to be consistent with import of FA3/FA4.

File: megatron/rl/rl_utils.py [hunk 2]
```diff
@@ -751,6 +757,136 @@ def selective_log_softmax(logits, index):
...
+
+
+def _build_cp_scatter_cache(
+    packed_seq_params: PackedSeqParams, seq_len: int, cp_group: torch.distributed.ProcessGroup
+) -> _CPScatterCache:
+    """Derive the CP partition for one bin: annotated params copy + partition indices."""
+    assert tex is not None, (
+        "Context-parallel RL logprobs require Transformer Engine to compute THD indices."
+    )
+    cp_rank = torch.distributed.get_rank(cp_group)
```
Comment (yobibyte) [inline megatron/rl/rl_utils.py:815]: Can we use raw torch get_rank() or we need to use the core's parallel_state for that?

File: megatron/rl/rl_utils.py [hunk 3]
```diff
@@ -751,6 +757,136 @@ def selective_log_softmax(logits, index):
...
+    cp_size = torch.distributed.get_world_size(cp_group)
+
+    assert seq_len % (2 * cp_size) == 0, (
+        f"Sequence length {seq_len} must be divisible by 2*context_parallel_size."
```
Comment (yobibyte) [inline megatron/rl/rl_utils.py:821]: Would be useful to explain why this is the case.

File: megatron/rl/rl_utils.py [hunk 4]
```diff
@@ -751,6 +757,136 @@ def selective_log_softmax(logits, index):
...
+    )
+
+    # Shallow-copy so we do not mutate the main sequence params.
+    cp_packed_seq_params = copy.copy(packed_seq_params)
+    vars(cp_packed_seq_params).pop('_rl_cp_scatter', None)
```
Comment (yobibyte) [inline megatron/rl/rl_utils.py]: Why are you doing this?

File: train_rl.py [hunk 5]
```diff
@@ -275,7 +271,9 @@ def forward_step(data_iterator, model: GPTModel, loss_only: bool = False):
                 max_seqlen_q=tokens.shape[1],
                 max_seqlen_kv=tokens.shape[1],
                 total_tokens=tokens.shape[1],
+                pad_between_seqs=False,
             )
+            packed_seq_params.seq_idx = None
```
Comment (yobibyte) [inline train_rl.py]: PackedSeqParams does not have seq_idx field.

File: megatron/rl/rl_utils.py [hunk 6]
```diff
@@ -751,6 +757,136 @@ def selective_log_softmax(logits, index):
...
 
+@dataclass
+class _CPScatterCache:
+    """Per-bin CP partition artifacts, derived once from (packed_seq_params, cp_group)."""
+
+    seq_len: int
+    cp_group: torch.distributed.ProcessGroup
+    cp_packed_seq_params: PackedSeqParams
+    partition_index: torch.Tensor
+    inverse_gather_perm: torch.Tensor
```
Comment (yobibyte) [inline megatron/rl/rl_utils.py:805]: This needs the docstring.

File: megatron/rl/rl_utils.py [hunk 7]
```diff
@@ -751,6 +757,136 @@ def selective_log_softmax(logits, index):
...
+    seq_len = tokens.shape[1]
+    # The partition depends only on the bin layout and CP topology. Derive it once and memoize.
+    cp_scatter = getattr(packed_seq_params, '_rl_cp_scatter', None)
+    if (
+        cp_scatter is None
+        or cp_scatter.cp_group is not cp_group
+        or cp_scatter.seq_len != seq_len
+    ):
+        cp_scatter = _build_cp_scatter_cache(packed_seq_params, seq_len, cp_group)
+        packed_seq_params._rl_cp_scatter = cp_scatter
```
Comment (yobibyte) [inline megatron/rl/rl_utils.py]: There is no field in the dataclass for this. I do not like this hidden interface. We should make this a part of the dataclass for people to know we are using it like that.

File: megatron/rl/rl_utils.py [hunk 8]
```diff
@@ -751,6 +757,136 @@ def selective_log_softmax(logits, index):
...
+        (local_tokens, local_position_ids, cp_packed_seq_params, local_labels, cp_scatter)
+        where every local tensor has sequence length `seq_len // cp_size` and `cp_scatter`
+        is this bin's _CPScatterCache.
+    """
+    seq_len = tokens.shape[1]
+    # The partition depends only on the bin layout and CP topology. Derive it once and memoize.
+    cp_scatter = getattr(packed_seq_params, '_rl_cp_scatter', None)
+    if (
+        cp_scatter is None
+        or cp_scatter.cp_group is not cp_group
```
Comment (yobibyte) [inline megatron/rl/rl_utils.py]: How is this condition possible?

File: megatron/rl/rl_utils.py [hunk 9]
```diff
@@ -751,6 +757,136 @@ def selective_log_softmax(logits, index):
...
+    return (
+        local_tokens,
+        local_position_ids,
+        cp_scatter.cp_packed_seq_params,
+        local_labels,
+        cp_scatter,
+    )
+
+
+def _gather_logprobs_context_parallel(
```
Comment (yobibyte) [inline megatron/rl/rl_utils.py]: We are using this function only once and it is three lines, do we really need a helper for it?

File: megatron/rl/rl_utils.py [hunk 10]
```diff
@@ -767,8 +903,7 @@ def get_logprobs(model, tokens, position_ids, no_grad=False, sequence_packing=Fa
             Required when packed_seq_params is provided to avoid CPU-GPU synchronization.
 
     Returns:
-        Logprobs of input sequences.
-
+        Logprobs of input sequences  [batch, seq_len - 1].
```
Comment (yobibyte) [inline megatron/rl/rl_utils.py]: These are not logprobs of input sequences, those are logprobs of outputs for those input sequences.

File: megatron/rl/rl_utils.py [hunk 11]
```diff
@@ -793,8 +928,13 @@ def get_logprobs(model, tokens, position_ids, no_grad=False, sequence_packing=Fa
                 max_seqlen_q=tokens.shape[1],
                 max_seqlen_kv=tokens.shape[1],
                 total_tokens=tokens.shape[1],
+                pad_between_seqs=False,
             )
+            packed_seq_params.seq_idx = None
```
Comment (yobibyte) [inline megatron/rl/rl_utils.py]: There is no such a field in PackedSeqParams

File: megatron/rl/rl_utils.py [hunk 12]
```diff
@@ -807,28 +947,46 @@ def get_logprobs(model, tokens, position_ids, no_grad=False, sequence_packing=Fa
...
         if not is_pp_last_stage(pp_group):
             return logits_or_hidden_states
-        else:
-            logits = logits_or_hidden_states
-            with nvtx_range("rl/log-softmax", time=True):
+
+        logits = logits_or_hidden_states
+        with nvtx_range("rl/log-softmax", time=True):
+            if cp_size > 1:
+                # Compute local logprobs then gather the full sequence.
```
Comment (yobibyte) [inline megatron/rl/rl_utils.py]: Please remove this useless comment.

File: megatron/rl/sequence_packing_utils.py [hunk 13]
```diff
@@ -620,12 +662,40 @@ def compute_packed_inference_logprobs_stats(
...
 
 class SequencePacker:
-    """Packs multiple sequences into bins to minimize padding and improve GPU utilization."""
+    """Packs multiple sequences into bins to minimize padding and improve GPU utilization.
+
+    Args:
+        bin_size: Size of each bin (padded sequence length).
+        pad_token: Padding token id.
+        max_sequences_per_bin: Maximum number of sequences per bin.
+        seq_length_multiple: Round every sequence's reserved footprint in the bin up
```
Comment (yobibyte) [inline megatron/rl/sequence_packing_utils.py]: "reserved footprint" is unclear

File: megatron/rl/sequence_packing_utils.py [hunk 14]
```diff
@@ -620,12 +662,40 @@ def compute_packed_inference_logprobs_stats(
...
+        assert bin_size % seq_length_multiple == 0, (
+            f"bin_size ({bin_size}) must be divisible by seq_length_multiple "
+            f"({seq_length_multiple})"
+        )
         self.bin_size = bin_size
         self.pad_token = pad_token
         self.max_sequences_per_bin = max_sequences_per_bin
+        self.seq_length_multiple = seq_length_multiple
+
+    def _reserved_footprint(self, length: int) -> int:
```
Comment (yobibyte) [inline megatron/rl/sequence_packing_utils.py]: Do we really need a function for one call we use only inside of a single method?

**Review by yobibyte — CHANGES_REQUESTED** [review]
Hey! Thanks for the PR, I left some comments mostly re code layout/readability.

**Comment by yobibyte** [comment]
Also, I did not read the tests code.

File: megatron/rl/rl_utils.py (hunk 2 above)
Comment (tdene) [reply megatron/rl/rl_utils.py:815]: The current guidance in Megatron LM is to use this style and avoid overusing `parallel_state`. I think they want this code style to avoid confusion where there's multiple parallelism setups, and `parallel_state` only reflects the current one being used, not necessarily the correct one. Like if a model exists in 2 parallelisms, you don't want to refer to `parallel_state` because it'll talk about the current live parallelism, but you may be asking questions about either the live or the suspended parallelisms.

File: megatron/rl/rl_utils.py (hunk 1 above)
Comment (tdene) [reply megatron/rl/rl_utils.py]: Absolutely right, fixed.

File: megatron/rl/rl_utils.py (hunk 3 above)
Comment (tdene) [reply megatron/rl/rl_utils.py:821]: I added a comment, what do you think?

File: megatron/rl/rl_utils.py (hunk 4 above)
Comment (tdene) [reply megatron/rl/rl_utils.py]: Because `_rl_cp_scatter` is a private attribute representing stale cached data that has no business being in the copy. No one would use it today, but I don't want future code to end up using stale data.

However, in addressing your other comments, I've just reworked this whole approach. These lines look a lot better now.

File: train_rl.py (hunk 5 above)
Comment (tdene) [reply train_rl.py]: It does; it's used to pre-compute some information and store it. If `seq_idx` ends up being not `None`, the packed kernel gets used. So to use the unpacked kernel we have to manually clear out this pre-computed cache.

I've gone ahead and changed the code so that we don't pre-compute `seq_idx` to begin with, instead of clearing it out after the fact.

File: megatron/rl/rl_utils.py (hunk 6 above)
Comment (tdene) [reply megatron/rl/rl_utils.py:805]: Done

File: megatron/rl/rl_utils.py (hunk 7 above)
Comment (tdene) [reply megatron/rl/rl_utils.py]: Fair. I copied it this way from the fork, and didn't clean it up because I didn't want to trigger more CODEOWNERS.

Fixed.

File: megatron/rl/rl_utils.py (hunk 8 above)
Comment (tdene) [reply megatron/rl/rl_utils.py]: It's not possible today, but this is guarding the misuse of cached information. It becomes possible as soon as we end up computing logprobs from two separate model through the same path for whatever reason.

I demoted this to an assert, fair?

File: megatron/rl/rl_utils.py (hunk 9 above)
Comment (tdene) [reply megatron/rl/rl_utils.py]: We do not need it. This function used to be a lot longer in the fork before I ported it over. Fixed.

File: megatron/rl/rl_utils.py (hunk 10 above)
Comment (tdene) [reply megatron/rl/rl_utils.py]: You're right, this is wrong. Reworded.

I don't like "logprobs of outputs", because outputs is overloaded and readers may think these are generation-only logprobs.

File: megatron/rl/rl_utils.py (hunk 11 above)
Comment (tdene) [reply megatron/rl/rl_utils.py]: It does; it's used to pre-compute some information and store it. If `seq_idx` ends up being not `None`, the packed kernel gets used. So to use the unpacked kernel we have to manually clear out this pre-computed cache.

I've gone ahead and changed the code so that we don't pre-compute `seq_idx` to begin with, instead of clearing it out after the fact.

File: megatron/rl/rl_utils.py (hunk 12 above)
Comment (tdene) [reply megatron/rl/rl_utils.py]: Done

File: megatron/rl/sequence_packing_utils.py (hunk 13 above)
Comment (tdene) [reply megatron/rl/sequence_packing_utils.py]: Reworded!

File: megatron/rl/sequence_packing_utils.py (hunk 14 above)
Comment (tdene) [reply megatron/rl/sequence_packing_utils.py]: There were two call sites, but yeah, this is bad. Fixed.

File: megatron/rl/rl_utils.py (hunk 8 above)
Comment (yobibyte) [reply megatron/rl/rl_utils.py]: Yeah, I like explicit better, thanks

File: megatron/rl/rl_utils.py (hunk 7 above)
Comment (yobibyte) [reply megatron/rl/rl_utils.py]: Thanks, we should not make the code shittier to avoid involving more reviewers.

File: train_rl.py (hunk 5 above)
Comment (yobibyte) [reply train_rl.py]: My bad!

File: megatron/rl/rl_utils.py (hunk 4 above)
Comment (yobibyte) [reply megatron/rl/rl_utils.py]: Cool, looks much better now. I do not like PackedSeqParams name as it is not really Params anymore, but that's is probably too late to change and is not the PR for it.

**Review by yobibyte — APPROVED** [review]

**Review by kvareddy — APPROVED** [review]

**Review by ericharper — APPROVED** [review]
